4단계 - 수강신청(요구사항 변경), 3단계 - 수강신청(DB 적용) 피드백 적용

mitsui3point opened the pull request:
안녕하세요 리뷰어님, 

이번미션에서 이전 table 을 보존하면서 code를 점진적으로 리팩토링하기 위해, 
Tobe 를 prefix 로 한 code 들과 table 을 생성했고,
step3 에서 수행했던 code 들과 table 은 기존방식대로 동작하도록 유지해두었습니다.

추가적으로 dbms 를 mapping 하게 되면서 getter 를 사용하는 빈도가 많이 는것 같습니다. 
어느 부분은 사용이 불가피하여 사용해야 하고,
어느부분은 남용하면 안되는 부분인지가 개념이 아직 모호한것 같아서,
최대한 dbms 와 연관된 domain 인 경우를 제외하고 getter 를 지양했습니다. 

이 부분(getter)은 어떤 방향으로 가닥을 잡아야 할지가 고민이 됩니다. 혹시 좋은 가이드나 기준이 있다면, 코멘트 남겨주시면 감사하겠습니다🙇

이번리뷰도 잘부탁드리겠습니다.🙇

mitsui3point committed:
refactor(lms): merge paid_session, free_session table to session table

refactor(lms): add session category field

refactor(lms): remove toParameters()

mitsui3point committed:
docs(lms): add README_LMS_CHANGEREQ TODO

mitsui3point committed:
docs(lms): modify README_LMS_CHANGEREQ TODO

mitsui3point committed:
docs(lms): modify README_LMS_CHANGEREQ TODO

mitsui3point committed:
docs(lms): modify README_LMS_CHANGEREQ TODO

mitsui3point committed:
docs(lms): modify README_LMS_CHANGEREQ TODO

mitsui3point committed:
feat(lms): add TobeStudent
add ApprovedStatus
add SelectedStatus

test(lms): add TobeStudentTest

feat(lms): remove Student toParameter() method
add Student getter methdods
modify JdbcStudentRepository toParameter -> getter method

test(lms): remove StudentTets toParameter() method
add StudentTest getter methdods

feat(lms): modify README_LMS_CHANGEREQ
modify TobeStudent(Student 수정) 추가
modify SelectedStatus field 추가
modify ApprovedStatus field 추가

mitsui3point committed:
feat(lms): add Instructor

test(lms): add InstructorTest

feat(lms): modify README_LMS_CHANGEREQ
modify Instructor 추가

mitsui3point committed:
feat(lms): modify Instructor constructor

test(lms): modify InstructorTest constructor

mitsui3point committed:
feat(lms): modify ImageType toImageType method

feat(lms): modify Instructor equal and hashcode method

test(lms): add CoverImageTest getter test

mitsui3point committed:
feat(lms): modify TobeCoverImage

test(lms): modify TobeCoverImageTest

docs(lms): modify README_LMS_CHANGEREQ TobeCoverImages 추가

mitsui3point committed:
docs(lms): modify README_LMS_CHANGEREQ TobeCoverImages 추가

mitsui3point committed:
fix(lms): remove unused Session constructor

mitsui3point committed:
feat(lms): add TobeSession
add TobeFreeSession
add TobePaidSession
add ProcessEndedException
add ProcessStatus
add RecruitmentStatus

test(lms): modify TobeSessionTest

docs(lms): modify README_LMS_CHANGEREQ TobeSession(TobeFreeSession, TobePaidSession) 생성

mitsui3point committed:
fix(lms): modify TobeStudent constructor signiture, equal and hashcode method

test(lms): add TobeFreeSessionTest fixture
add TobePaidSessionTest fixture
modify TobeStudentTest constructor tests

mitsui3point committed:
feat(lms): add TobeFreeSession register method
add TobePaidSession register method
add PaymentStudentNsUserNotMatchException
add RecruitmentClosedException
add RecruitmentStatus

test(lms): modify TobeSessionTest

docs(lms): modify README_LMS_CHANGEREQ 수강신청은 RecruitmentStatus(OPEN) 일때 가능하다
add  README_LMS_CHANGEREQ TobeCoverImages 추가

mitsui3point committed:
feat(lms): add TobeStudents

test(lms): add TobeStudentsTest

feat(lms): modify unused code block (TobePaidSession, TobeStudent)

feat(lms): modify sort code (TobePaidSession, TobeStudent)

docs(lms): modify README_LMS_CHANGEREQ TobeSessions 생성

mitsui3point committed:
fix(lms): modify TobeSession TobeCoverImage -> List<TobeCoverImage> field

test(lms): add TobeStudentsTest

mitsui3point committed:
fix(lms): add TobeSession instructorId field
modify TobeSession constructor signiture
modify TobeFreeSession constructor signiture
modify TobePaidSession constructor signiture
add Instructor getId

test(lms): add Instructor getter
modify TobeSession constructor signiture 관련 test

mitsui3point committed:
feat(lms): add TobeCoverImages
modify TobeSession field List<TobeCoverImage> -> TobeCoverImages

test(lms): add TobeCoverImagesTest

docs(lms): modify README_LMS_CHANGEREQ TobeCoverImages 추가

mitsui3point committed:
feat(lms): add TobeSession field TobeStudents (merge TobeFreeSession field TobeStudents, TobePaidSession field TobeStudents)

test(lms): add TobeSessionTest approveAll, deniedAll

feat(lms): add TobeStudents each

test(lms): add TobeStudentsTest each

feat(lms): add TobeStudent approved, denied

test(lms): add TobeStudentTest approved, denied, approved_승인처리패스, denied_취소처리패스

docs(lms): modify README_LMS_CHANGEREQ 강사가 수강 승인한다, 강사가 수강 취소한다

mitsui3point committed:
feat(lms): add schema.sql
tobe_session
cover_image
instructor
tobe_student

docs(lms): modify README_LMS_CHANGEREQ
status -> process_status column 변경
recruitment_status column 추가
cover_image table 추가
instructor table 추가
selected_status column 추가
approved_status column 추가

mitsui3point committed:
feat(lms): add TobeSessions

test(lms): add TobeSessionsTest

feat(lms): add Course TobeSessions field

test(lms): add CourseTest addTobeSession

mitsui3point committed:
feat(lms): add TobeSessionRepository, TobeJdbcSessionRepository

test(lms): add TobeSessionRepositoryTest

feat(lms): add TobeStudentRepository, TobeJdbcStudentRepository

test(lms): add TobeStudentRepositoryTest

mitsui3point committed:
fix(lms): modify TobeSessionTest getter failed test

mitsui3point committed:
feat(lms): add TobeCoverImageRepository, TobeJdbcCoverImageRepository

test(lms): add TobeCoverImageRepositoryTest

feat(lms): add TobeCoverImages add, size, getter

test(lms): add TobeCoverImagesTest add, size, getter

feat(lms): add TobeCoverImage sessionId field
remove creatorId field
modify constructor signiture

mitsui3point committed:
feat(lms): add TobeInstructorRepository, TobeJdbcInstructorRepository

test(lms): add TobeInstructorRepositoryTest

feat(lms): remove Instructor field creatorId
modify constructor signiture

mitsui3point committed:
feat(lms): add LocalDateTimeFormatter

dhmin5693 commented:
getter 사용은 나쁘지 않아요.
단지 getter를 사용하여 데이터를 꺼내오고 무언가를 하는 습관을 지양하자는 것이죠.
데이터를 꺼내다 주는게 아니라 객체에 메시지를 전달하는 방향이라고 이해하시면 돼요.

예를 들면 일급 컬렉션이 있다고 해볼게요.

```java
class Cars {
    private List<Car> items;
   ...
}

// 다른 클래스의 메소드
public void add(Car car) {
    cars.getItems().add(car); // getter를 호출한 다음 수행했기 때문에 add에 대한 책임을 완전히 가져가지 않음
    cars.add(car); // `Cars` class에서 add에 대한 책임까지 가져가게 됨
}
```

간단한 예제였지만 `Cars`에 저장된 데이터를 getter로 꺼내기만 했을 뿐 그 객체가 가져야 할 책임이 덜 부여되어 있다는걸 느끼시면 될 것 같아요.

dhmin5693 commented on src/main/resources/schema.sql:
```diff
@@ -51,36 +51,80 @@ create table delete_history (
+
+-- tobe
+
+create table tobe_session (
```
## 이 테이블의 필요 여부

이번 단계는 3가지의 추가 요구사항을 반영해야 합니다. 만약 수강신청이 4단계가 아니라 5단계로 끝나며 거기서도 요구사항을 3개쯤 더 추가한다면 어떻게 되나요?

tobe_tobe_session을 또 만들수는 없을거라고 생각해요. 
현업에서도 현실적으로 요구사항의 변경이 굉장히 잦고 이럴 때마다 새 테이블을 만들어서 매번 migration을 수행하지는 못합니다. 
테이블의 중요도에 따라 어디까지 영향이 미칠지 예측하기 쉽지 않기 때문입니다. 
따라서 저는 tobe_session은 불필요하다고 생각합니다.

아래는 제 의견의 근거입니다. 이번 단계의 요구사항이에요.

### 핵심 학습 목표
- DB 테이블이 변경될 때도 스트랭글러 패턴을 적용해 점진적인 리팩터링을 연습한다.

### 프로그래밍 요구사항
- 리팩터링할 때 컴파일 에러와 기존의 단위 테스트의 실패를 최소화하면서 점진적인 리팩터링이 가능하도록 한다.
- DB 테이블에 데이터가 존재한다는 가정하에 리팩터링해야 한다.
    - 즉, 기존에 쌓인 데이터를 제거하지 않은 상태로 리팩터링 해야 한다.
        - (리뷰어 덧붙임) 저는 여기에서 기존 테이블에 컬럼을 추가하여 진행해야 한다고 이해했습니다.

게다가 실제로 사용하고 있지는 않네요. TobeSessionRepository 등은 test에서만 호출되고 있습니다.
tobe_session 테이블과 부속 코드는 전부 삭제해주세요.

## 만약에 꼭 사용해야 한다면?

이름은 조금 더 신중하게 짓는게 좋겠습니다. 이렇게 명명한 이유는 알겠으나 테이블명은 한 번 정해지면 바꾸기 어렵습니다. 시간이 지나면 이 테이블의 이름이 왜 tobe이고 언제까지 tobe인지 알 수 없어요. 

그리고 위에도 썼지만 또 추가했을 때 tobe_tobe로 짓기도 조금 그렇습니다.

mitsui3point replied:
아 tobe_tobe_..의 연속이 될거라는 생각까지 미치지 못했네요..

말씀하신 부분에서 오해가 생긴게 정확히 맞는것 같습니다.

저 부분에서 저는 단순하게 
1. 새 테이블을 만든다.
2. 새 기능구현 완료시 migration 을 한다.

라고 생각하고 무작정 달려들었던게 문제였던것 같습니다.

제가 이전 미션들 자동차경주때와 사다리타기때에도
미션의 요구사항을 오분석하는경우가
스텝중에 꼭 한번씩 있었습니다.

다른 참가자분들의 코드들을 가끔 보다보면
요구사항을 오분석하여 
거의 새로 갈아엎어야 하는 수준까지는 도달하지 않는것 같은데
유독 제가 꼭 한번씩 그런 경험을 하는것 같아 고민이 많습니다.
(실제 현업에서도 종종 '내가말한건 이게 아닌데..' 라는말을 듣는 경우가 있습니다..)

리뷰어님은 이런 상황이 만약 있다고 할때,
다음 대처를 어떻게 생각하고 계신지 궁금합니다.
(난제같긴하지만 유독 제가 그런 경험을 많이하는것 같아 고민이 되는 주제입니다..😭)

dhmin5693 replied:
음... 어려운 문제입니다만 본인이 오해를 하셨다면 적당하게 잘 넘어가는게 좋은 방법 아닐까요?
저도 항상 커뮤니케이션이 가장 어렵더라고요.

그리고 미션 자체가 큰 틀 안에서만 맞으면 괜찮긴 해서 본인이 생각하신 바 있으면
"저는 ~ 이렇게 생각했는데 혹시 오해가 있었던 건가요?" 처럼 의견을 교환하시는 것도 괜찮아요.

일반적으로 리뷰는 학습을 지원해드리기 위한 절차이지만 학습자와 리뷰어가 소통하는 자리이기도 해요.
학습자가 작성한 코드의 동작은 이해할 수 있어도 의도는 100% 이해할 수 없기도 합니다.
이 과정 뿐만이 아니라 작성한 코드에 대해 적당한 이유가 있고 그게 논리적으로 납득 가능하다 느껴진다면 꼭 언급해주시는게 좋아요.
현업에서도 마찬가지이구요.

mitsui3point replied:
감사드립니다. 불안한 마음에 시작된 질문이었는데, 답변을 받고나니 마음이 한결 가벼워졌습니다.

생각해보니 리뷰어님이 남겨주신 코멘트로도 코드리뷰 라는 커뮤니케이션을 배워가는것 같아요. 
제가 학습자입장이다보니 해내기 급급하다 보니 커뮤니케이션의 역량도 챙겨가고 있는걸 새삼 느끼게 되었습니다.

이렇게 좀 모호하고 추상적인 질문에도 도움되는 답글 남겨주셔서 감사합니다. 하나하나 정말 잘 배우고있어요.🙇‍♂️

mitsui3point replied:
감사드립니다. 불안한 마음에 시작된 질문이었는데, 답변을 받고나니 마음이 한결 가벼워졌습니다.

생각해보니 리뷰어님이 남겨주신 코멘트로도 코드리뷰 라는 커뮤니케이션을 배워가는것 같아요. 
제가 학습자입장이다보니 해내기 급급하다 보니 커뮤니케이션의 역량도 챙겨가고 있는걸 새삼 느끼게 되었습니다.

이렇게 좀 모호하고 추상적인 질문에도 도움되는 답글 남겨주셔서 감사합니다. 하나하나 정말 잘 배우고있어요.🙇‍♂️

dhmin5693 commented on src/main/java/nextstep/courses/infrastructure/JdbcSessionRepository.java:
```diff
@@ -1,27 +1,135 @@
+        ImageSize imageSize = coverImage.getImageSize();
+        ImageType imageType = coverImage.getImageType();
+        Status status = session.getStatus();
+        return jdbcTemplate.update(sql,
+                session.getId(), session.getCourseId(), category.name(), dateRange.getStartAt(), dateRange.getEndAt(),
+                imageFileSize.getSize(), imageType.name(), imageSize.getWidth(), imageSize.getHeight(), status.name(),
+                session.getMaxRegisterCount(), session.getAmount(), session.getCreatorId(), session.getCreatedAt()
```
이 부분때문에 getter에 대한 질문을 남겨주신 것 같은데 위에 남겨놓은 답변을 확인해주세요.

저는 괜찮은 부분이라고 보입니다.
다만 getter가 너무 많은건 이 클래스에 필드가 많기 때문이고 요구사항을 담기 위해서라면 어느 정도는 감안해야 하는 점이 아닐까 싶어요.

mitsui3point replied:
위의 답변 잘 보았습니다

말씀해주신것처럼 
> getter 사용은 나쁘지 않아요.
> 단지 getter를 사용하여 데이터를 꺼내오고 무언가를 하는 습관을 지양하자는 것이죠.
> 데이터를 꺼내다 주는게 아니라 객체에 메시지를 전달하는 방향이라고 이해하시면 돼요.

객체에 메시지를 전달할 방법이 충분히 있음에도 불구하고
무분별하게 getter 를 사용하는것을 지양해야 한다는 말씀으로 이해했습니다.

이번 과정이 그것을 해보자는 취지에서 getter 를 제한적으로 사용하다보니
너무 한쪽방향으로 치우쳐서 생각한것 같습니다.
예시를 들어주셔서 한방에 이해했습니다.! 감사합니다🙇

dhmin5693 reviewed (changes requested):
안녕하세요.

다 괜찮게 잘 해주셨는데 학습 목표의 이해가 저랑 약간 다른 것 같습니다.
남겨드린 리뷰 참고하여 수정 후 다시 요청해주세요.

이해되지 않는 내용 있으시면 언제든지 DM으로 질문해주셔도 괜찮습니다.

mitsui3point committed:
docs(lms): add README_LMS_CHANGEREQ_PRE_FEEDBACK
modify README_LMS_CHANGEREQ TODO

mitsui3point committed:
feat(lms): remove Course TobeSession field

test(lms): remove CourseTest addTobeSession

mitsui3point committed:
feat(lms): modify Student (TobeStudent -> Student)

test(lms): modify Student (TobeStudentTest -> StudentTest)

docs(lms): modify README_LMS_CHANGEREQ Student 수정

mitsui3point committed:
feat(lms): add Instructor

test(lms): add InstructorTest

docs(lms): modify README_LMS_CHANGEREQ Instructor 추가

mitsui3point committed:
feat(lms): add CoverImage

test(lms): add CoverImageTest

docs(lms): modify README_LMS_CHANGEREQ CoverImage 수정

mitsui3point committed:
test(lms): modify TobeInstructorRepository -> InstructorRepository

mitsui3point committed:
test(lms): modify TobeCoverImagesRepository -> CoverImagesRepository

mitsui3point committed:
feat(lms): modify TobeCoverImagesRepository -> CoverImagesRepository

test(lms): modify TobeCoverImagesRepositoryTest -> CoverImagesRepositoryTest

mitsui3point committed:
feat(lms): modify TobeSessionRepository -> SessionRepository
modify TobePaidSessionRepository -> PaidSessionRepository
modify TobeFreeSessionRepository -> FreeSessionRepository
modify TobeStudentsRepostory -> StudentsRepository
modify TobeStudents -> Students
add tobe package exceptions

test(lms): modify TobeCoverImagesRepositoryTest -> CoverImagesRepositoryTest
modify TobePaidSessionRepositoryTest -> PaidSessionRepositoryTest
modify TobeFreeSessionRepositoryTest -> FreeSessionRepositoryTest
modify TobeStudentsRepostoryTest -> StudentsRepositoryTest
modify TobeStudentsTest -> StudentsTest

docs(lms): modify README_LMS_CHANGEREQ
Sessions 변경, CoverImages 추가,
수강신청은 RecruitmentStatus(OPEN) 일때 가능하다
강의는 하나 이상의 커버 이미지를 가질 수 있다.
학생은 수강신청을 할 수 있다.
강사가 선발된 인원들을 수강 승인한다
강사가 선발되지 않은 인원들을 수강 취소한다

mitsui3point committed:
feat(lms): remove tobe packages, tobe package 종속 imports

feat(lms): schema.sql tobe_* tables DDL

mitsui3point committed:
feat(lms): cleanup code, test code variable 필드화

dhmin5693 reviewed (approved):
반영 잘 해주셨네요 😄 
마지막 미션까지 정말 수고 많으셨습니다.

이 과정을 통해서 많은 것을 배워가셨으면 하는 마음이 있네요.
멀리서나마 항상 응원하겠습니다.
